db/src/tx.rs

```diff
@@ -245,157 +348,133 @@ impl<'conn, 'a> Tx<'conn, 'a> {
+                                // :db/id is optional; if it's not given, we generate a special internal tempid
+                                // to use for upserting.  This tempid will not be reported in the TxReport.
+                                let db_id: Option<entmod::EntidOrLookupRefOrTempId> = mentat_tx_parser::remove_db_id(&mut map_notation)?;
+                                let mut dangling = db_id.is_none();
+                                let db_id: entmod::EntidOrLookupRefOrTempId = db_id.unwrap_or_else(|| in_process.allocate_mentat_id());
+
+                                // We're nested, so we want to ensure we're not creating "dangling"
+                                // entities that can't be reached.  If we're :db/isComponent, then this
+                                // is not dangling.  Otherwise, the resulting map needs to have a
+                                // :db/unique :db.unique/identity [a v] pair, so that it's reachable.
+                                // Per http://docs.datomic.com/transactions.html: "Either the reference
+                                // to the nested map must be a component attribute, or the nested map
+                                // must include a unique attribute. This constraint prevents the
+                                // accidental creation of easily-orphaned entities that have no identity
+                                // or relation to other entities."
+                                if attribute.component {
                                     dangling = false;
                                 }
 
-                                deque.push_front(Entity::AddOrRetract {
-                                    op: OpType::Add,
-                                    e: db_id.clone(),
-                                    a: entmod::Entid::Entid(inner_entid),
-                                    v: inner_v,
-                                });
-                            }
+                                for (inner_a, inner_v) in map_notation {
+                                    if inner_a.is_backward() {
+                                        // We definitely have a reference.  The reference might be
+                                        // dangling (a bare entid, for example), but we don't yet
+                                        // support nested maps and reverse notation simultaneously
+                                        // (i.e., we don't accept {:reverse/_attribute {:nested map}})
+                                        // so we don't need to check that the nested map reference isn't
+                                        // danglign.
+                                        dangling = false;
+
+                                        let reversed_e = in_process.entity_v_into_term_e(inner_v, &inner_a)?;
+                                        let (reversed_a, _reversed_attribute) = in_process.entity_a_into_term_a(inner_a.to_reversed().unwrap())?;
```

Comment:
You reverse `inner_a` immediately inside `entity_v_into_term_e`, and again here, and you `unwrap` here reliant on `is_backward` above.

I think you might be better off having a function `unreverse` alongside `is_backward` and `to_reversed` that returns  `Option<Entid>`. Use that with `if let` instead of `is_backward`. That gives you the reversed ident immediately, does the work only once, and gets rid of the `unwrap` here.